doc/source/data/doc_code/custom_datasource.py

```diff
@@ -0,0 +1,115 @@
+# flake8: noqa
+
+# fmt: off
+# __read_single_query_start__
+from ray.data.block import Block
+
+def _read_single_query(uri, database, collection, query, schema, kwargs) -> Block:
+    import pymongo
+    from pymongoarrow.api import aggregate_arrow_all
+
+    client = pymongo.MongoClient(uri)
+    return aggregate_arrow_all(
+        client[database][collection], query, schema=schema, **kwargs
+    )
+# __read_single_query_end__
+# fmt: on
+
+# fmt: off
+# __mongo_datasource_reader_start__
+from typing import Any, Dict, List, Optional
+from ray.data.datasource.datasource import Datasource, Reader, ReadTask
+from ray.data.block import BlockMetadata
+
+class _MongoDatasourceReader(Reader):
```

Comment:
I feel we should add more comments in this file. For example, we we should say something like constructor args are passed in from `read_datasource()`. What's `parallelism` of `get_read_tasks` and how to use it.